### DISCUSSION TITLE
如何为 umijs 编写插件？

### DISCUSSION THREAD

**Discussion by xXAvoraXx** [umijs/umi#11745] [General]
您好，我们有一个在数据库中保存路由和组件的结构。当用户访问路由时，我们会从数据库中提取该路由的组件的凭据，并决定是否进行渲染。我们想针对这种情况编写一个插件。我们该怎么做呢？

**Comment by fz6m** [comment]
如果是简单的路由权限校验，可以直接给应用 layout 套一层 auth context ，在这里面校验用户有没有这个路由的权限就可以。如果需要动态路由表，需要使用 [`patchClientRoutes`](https://umijs.org/docs/api/runtime-config#patchclientroutes-routes-) ，一个例子见：https://github.com/umijs/umi/discussions/9717#discussioncomment-4078053

**Reply by xXAvoraXx** [reply]
你能检查一下吗？如果用户注销并重新登录，则不会加载路由的组件。因此，用户必须刷新页面。我无法解决这个问题。
https://github.com/ant-design/ant-design-pro/discussions/10898

**Reply by xXAvoraXx** [reply]
![image](https://github.com/umijs/umi/assets/6988989/34929302-f5cc-4d1e-aa85-ccc51ef1d3b2)
我有两个表，分别是路由表和组件表。当用户到达相关路由时，我会从数据库中获取连接到该路由的组件列表。 但是，我如何编写代码，确定哪些组件将在该部分呈现或不呈现？

**Reply by fz6m** [reply]
在 umi 应用中，无论是登录还是注销，都必须自动刷新页面重载应用，否则权限和路由都不会更新。

通常情况下，会单独有一个页面用来登录，或者跳转到 sso 去登录，最后登录完毕都会刷新页面重载应用。

**Reply by xXAvoraXx** [reply]
如何确定哪些组件已渲染，哪些未渲染？当用户访问相关路由时，我会从网络应用程序请求该路由的组件列表（key-id）。

**Reply by xXAvoraXx** [reply]
https://github.com/ant-design/ant-design-pro/discussions/10898

**Reply by fz6m** [reply]
需要维护一个全部动态路由的映射表，用户会根据路由映射表添加有权限的路由，参考 https://github.com/umijs/umi/discussions/9717#discussioncomment-4078053

在上面你那个 issue 里描述的问题上下文过于复杂，目前已经提供了可以跑通的最小动态路由例子，你有什么自定义需求基于这个最小例子代码添加自己的逻辑即可。
